### DISCUSSION TITLE
Secret/Public key naming

### DISCUSSION THREAD

**Discussion by tcharding** [rust-bitcoin/rust-bitcoin#974] [Refactoring]
## TL;DR
Consider using the identifiers `seckey` and `pubkey` across the whole stack.

EDIT: Related to https://github.com/rust-bitcoin/rust-bitcoin/discussions/4950

## 

Keys are our bread and butter yet across our stack we have a confusing array of different identifiers. I would like to please discuss simplifying/unifying all the names.

### secp256k1

- `SecretKey`
- `PublicKey`

### bitcoin

- `PrivateKey`
- `PublicKey`
- `bip32::ExtendedPrivKey`
- `bip32::ExtendedPubKey`
- `XOnlyPublicKey`

### miniscript

- `DescriptorSecretKey`
- `DescriptorPublicKey`

## Assumptions

- We should use conventional Rust capitalization, if pubkey is a 'word' then it should appear as `Pubkey` not `PubKey` (and `pubkey` not `pub_key`) 

- Two things can have the same name and to distinguish them users add the module path i.e., `bitcoin::Pubkey` and `secp256k1::Pubkey`. This is totally acceptable and is the current best practice (as opposed to `BitcoinPubkey`).

## Candidate identifiers

- `Seckey` / `Pubkey`
- `ExtendedSeckey` / `ExtendedPubkey`
- `DescriptorSeckey` / `DescriptorPubkey`
- `XOnlyPubkey`

Explicitly off topic for this discussion please:
-  `XOnlyPubkey` vs `XonlyPubkey` (makes variables more ergonomic e.g., `xonly_pubkey`)

**Comment by apoelstra** [comment]
The discrepancy between privkey and seckey actually comes from upstream -- in libsecp a "secret key" is a raw 32-byte value, while in Bitcoin a "private key" is a (typically WIF-encoded) blob of data which also includes a network identifier and a compressedness byte. Similarly "xpriv" and "xpub" are widely-understood terms and if we chonged "private" to "secret" here it would likely be confusing.

Agree that we should unify capitalization at least.

**Reply by Kixunil** [reply]
Interesting, I've heard that "secret" is usually used for symmetric encryption and "private" for asymmetric. So in a way upstream may have unusual naming?

**Reply by apoelstra** [reply]
It might. I haven't heard this before, but on the other hand I almost never work with symmetric encryption (nor do the other authors of libsecp, I believe)

**Comment by tcharding** [comment]
Oh cool, thanks for the explanation. So what if anything that was (or wraps) a secp key used seckey/pubkey and anything that was (or wraps) a bitcoin key used privkey/pubkey? I'd like to also add docs to rust-secp/rust-bitcoin with your explanation above - I've wondered about this for 3 years and never got round to asking why :)

**Comment by apoelstra** [comment]
> So what if anything that was (or wraps) a secp key used seckey/pubkey and anything that was (or wraps) a bitcoin key used privkey/pubkey? 

This was my original intention, but then with Taproot suddenly we're using bare secp secret keys :)

**Reply by tcharding** [reply]
Drat and bother. Cheers man.
